### ISSUE TITLE
Training on Custom Dataset

### ISSUE THREAD

**Issue by rayush7** [bgshih/crnn#61]
@bgshih 
I am trying to train crnn on a synthetic dataset that I generated. Along with english alphabets and digits (0-9) it contains special characters like $,@,# etc. I was able to modify the utilities.lua to incorporate their ascii value and hence created the lmdb dataset. But when I am trying to train the model on the synthetic dataset then the predicted value are all the same. I am confused about where I am making the mistake.

![pic1](https://cloud.githubusercontent.com/assets/15009941/26449799/e81b9f52-4171-11e7-9f28-056a66391d0d.png)
![pic2](https://cloud.githubusercontent.com/assets/15009941/26449800/e81cf730-4171-11e7-8871-6f555363bbd2.png)

Also I want to know is it possible to load the weight from the pretrained crnn model (trained on vgg synthetic dataset) before I start training the model on a new dataset? How do I go about doing that?

**Comment by shoaibazam** [comment]
@bgshih 
Thanks for this package. I am using this repo for training on my own dataset. As there is code to create dataset in tool/create_dataset.py, but when I run this code I got nothing even after specifying all the parameters. I would like to request you to please help me in this matter. Thanks

**Comment by stalagmite7** [comment]
Hi @rayush7 , I have been trying to retain the network on my custom dataset as well, but the training process after loading a batch errors out at the forward pass (in training.lua in src/) with this message:
`
     terminate called after throwing an instance of 'std::invalid_argument'
     what():  index out of range
     Aborted (core dumped)
`


Any idea what's going on? Did you ever encounter such a problem?
@bgshih , suggestions?

**Comment by rayush7** [comment]
@stalagmite7 I think there is some problem with your lmdb dataset. Are you using this [script](https://github.com/bgshih/crnn/blob/master/tool/create_dataset.py) to generate your lmdb files?

**Comment by stalagmite7** [comment]
Yes, with one small change: instead of checking image validity thru opencv, I use the inbuilt imghdr package from python (since I am using a docker container to work on this, didn't want to have to rebuild the container with opencv):
`

def checkImageIsValid(imageBin):
    if imageBin is None:
        return False
    ret_val = imghdr.what(None, imageBin)
    if ret_val == 'png':
        return True
    return False
    """
    if imageBin is None:
        return False
    imageBuf = np.fromstring(imageBin, dtype=np.uint8)
    img = cv2.imdecode(imageBuf, cv2.IMREAD_GRAYSCALE)
    imgH, imgW = img.shape[0], img.shape[1]
    if imgH * imgW == 0:
        return False
    return True
    """


`

**Comment by stalagmite7** [comment]
Could it be that my ascii conversion is incorrect? Since the error it throws is an 'index out of range' error, perhaps I am trying to access indices for ascii characters that are out of the range of ascii?
This is the function I am using in src/utilities.lua:

function str2label(strs, maxLength)
            --[[ Convert a list of strings to integer label tensor (zero-padded).

            ARGS:
              - `strs`     : table, list of strings
              - `maxLength`: int, the second dimension of output label tensor

            RETURN:
              - `labels`   : tensor of shape [#(strs) x maxLength]
            ]]
            assert(type(strs) == 'table')

            function ascii2label(ascii)
        	print(ascii)
        	local label
        	if ascii >= 33 and ascii <= 47 then -- symbols '!' to '/' mapped to 1-14
                    label = ascii - 32
                
        	elseif ascii >= 48 and ascii <= 57 then -- '0'-'9' are mapped to 15-24
                    label = ascii - 47 + 14
        	
        	elseif ascii >= 58 and ascii <= 64 then --symbols ':' to '@' are mapped to 25-31
                    label = ascii - 57 + 24

                elseif ascii >= 65 and ascii <= 90 then -- 'A'-'Z' are mapped to 32-57
                    label = ascii - 64 + 31
         
                elseif ascii >= 97 and ascii <= 122 then -- 'a'-'z' are mapped to 58-83
                    label = ascii - 96 + 57
         
                elseif ascii >= 91 and ascii <= 96 then -- symbols '[' to '`' are mapped to 84-109
                    label = ascii - 90 + 83

                elseif ascii >= 123 and ascii <= 126 then --symbols '{' to '~' are mapped to 109-112
                    label = ascii - 122 + 109
                end
                return label
            end

            local nStrings = #strs
            local labels = torch.IntTensor(nStrings, maxLength):fill(0)
            for i, str in ipairs(strs) do
                for j = 1, string.len(str) do
                    local ascii = string.byte(str, j)
                    labels[i][j] = ascii2label(ascii)
                end
            end
            return labels
        end

Does anything seem immediately incorrect?

**Comment by rayush7** [comment]
The first "if condition" in ascii2label function seems incorrect. As when ascii value would be 47 then label would be 15 and not 14 (which you have mentioned in the comments). And in the last "elseif condition" you are mapping from 109-112 or 110-113?. Check it.
Apart from this not only you have to make changes in ascii2label function but also in [label2ascii](https://github.com/bgshih/crnn/blob/master/src/utilities.lua#L53) function, which is just the reverse conversion from labels to the ascii value.

**Comment by stalagmite7** [comment]
Yes, I figured out the ascii2label function error once I looked at it; didn't know of the label2ascii function though, thanks a ton for pointing it out!
This should work hopefully. :)

**Comment by shoaibazam** [comment]
Can this package be used for Chinese characters also. If so what change I have to make. Kindly guide me. Thanks.

**Comment by stalagmite7** [comment]
I made the corrections in the ascii2label and the lanel2ascii functions, but the training seems to go find through the ascii2label function, but breaks out with the same error during forward pass (and as far as I can tell, it isn't going into the label2ascii function before that, none of my prints in that function are seen). Is there any where else that the range of acceptable ascii characters is limited?
The training proceeds fine if my dataset doesn't have anything outside of the default range of alphanumeric characters

**Comment by LiangHao92** [comment]
@shoaibazam hello, have you successfully trained the model for chinese characters?

**Comment by Heisenberg0391** [comment]
@shoaibazam the model it self is language independent, it works well on my own chinese dataset
